### PR TITLE
extend SF dialect for File Format statements

### LINKED ISSUE #3773: Lack of "ALTER FILE FORMAT" statement and some Format Type Options for CSV, JSON and PARQUET types
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

1. In the snowflake dialect there's lack of "ALTER FILE FORMAT" statement for proper parsing SQL
[https://docs.snowflake.com/en/sql-reference/sql/alter-file-format.html#](url)
2.  For CSV type such formatTypeOptions are missed: 
```
ESCAPE = '<character>' | NONE
ESCAPE_UNENCLOSED_FIELD = '<character>' | NONE
```
EXAMPLE:

```
CREATE FILE FORMAT IF NOT EXISTS my_csv_format
              TYPE = CSV
              ESCAPE = '\\'
          ;
```

Error: Found unparsable section: 'CREATE  FILE FORMAT IF NOT EXISTS"

3. For PARQUET type 
`BINARY_AS_TEXT = TRUE | FALSE` 
formatTypeOption is missed

4. Also `NULL_IF=()` for JSON and CSV types cannot be parsed. It's needed to state `NULL_IF=('') `. However it's a allowable syntax to write `NULL_IF=()` [https://snowflakecommunity.force.com/s/article/how-to-update-file-format-with-nullif-to-remove-n-null-values-when-using-copy-into-location](url)

```
CREATE OR REPLACE FILE FORMAT UPDATED_FORMAT_NAME
TYPE = 'CSV'
COMPRESSION = 'NONE'
FIELD_DELIMITER ='|'
NULL_IF=()
```

### Expected Behaviour

SQL statements should be parsed without error

### Observed Behaviour

```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        create_file_format_segment:
[L:  1, P:  1]      |            keyword:                                          'CREATE'
[L:  1, P:  7]      |            whitespace:                                       ' '
[L:  1, P:  8]      |            keyword:                                          'FILE'
[L:  1, P: 12]      |            whitespace:                                       ' '
[L:  1, P: 13]      |            keyword:                                          'FORMAT'
[L:  1, P: 19]      |            whitespace:                                       ' '
[L:  1, P: 20]      |            keyword:                                          'IF'
[L:  1, P: 22]      |            whitespace:                                       ' '
[L:  1, P: 23]      |            keyword:                                          'NOT'
[L:  1, P: 26]      |            whitespace:                                       ' '
[L:  1, P: 27]      |            keyword:                                          'EXISTS'
[L:  1, P: 33]      |            whitespace:                                       ' '
[L:  1, P: 34]      |            object_reference:
[L:  1, P: 34]      |                naked_identifier:                             'my_file_format'
[L:  1, P: 48]      |            newline:                                          '\n'
[L:  2, P:  1]      |            whitespace:                                       '    '
[L:  2, P:  5]      |            csv_file_format_type_parameters:
[L:  2, P:  5]      |                keyword:                                      'TYPE'
[L:  2, P:  9]      |                whitespace:                                   ' '
[L:  2, P: 10]      |                comparison_operator:
[L:  2, P: 10]      |                    raw_comparison_operator:                  '='
[L:  2, P: 11]      |                whitespace:                                   ' '
[L:  2, P: 12]      |                file_type:                                    'CSV'
[L:  2, P: 15]      |                newline:                                      '\n'
[L:  3, P:  1]      |                whitespace:                                   '\t'
[L:  3, P:  2]      |        unparsable:                                           !! Expected: 'Nothing...'
[L:  3, P:  2]      |            raw:                                              'ESCAPE'
[L:  3, P:  8]      |            whitespace:                                       ' '
[L:  3, P:  9]      |            raw:                                              '='
[L:  3, P: 10]      |            whitespace:                                       ' '
[L:  3, P: 11]      |            raw:                                              "'\\\\'"
[L:  3, P: 15]      |    statement_terminator:                                     ';'
[L:  3, P: 16]      |    newline:                                                  '\n'
[L:  4, P:  1]      |    newline:                                                  '\n'
[L:  5, P:  1]      |    statement:
[L:  5, P:  1]      |        create_file_format_segment:
[L:  5, P:  1]      |            keyword:                                          'CREATE'
[L:  5, P:  7]      |            whitespace:                                       ' '
[L:  5, P:  8]      |            keyword:                                          'FILE'
[L:  5, P: 12]      |            whitespace:                                       ' '
[L:  5, P: 13]      |            keyword:                                          'FORMAT'
[L:  5, P: 19]      |            whitespace:                                       ' '
[L:  5, P: 20]      |            keyword:                                          'IF'
[L:  5, P: 22]      |            whitespace:                                       ' '
[L:  5, P: 23]      |            keyword:                                          'NOT'
[L:  5, P: 26]      |            whitespace:                                       ' '
[L:  5, P: 27]      |            keyword:                                          'EXISTS'
[L:  5, P: 33]      |            whitespace:                                       ' '
[L:  5, P: 34]      |            object_reference:
[L:  5, P: 34]      |                naked_identifier:                             'my_file_format'
[L:  5, P: 48]      |            newline:                                          '\n'
[L:  6, P:  1]      |            whitespace:                                       '    '
[L:  6, P:  5]      |            csv_file_format_type_parameters:
[L:  6, P:  5]      |                keyword:                                      'TYPE'
[L:  6, P:  9]      |                whitespace:                                   ' '
[L:  6, P: 10]      |                comparison_operator:
[L:  6, P: 10]      |                    raw_comparison_operator:                  '='
[L:  6, P: 11]      |                whitespace:                                   ' '
[L:  6, P: 12]      |                file_type:                                    'CSV'
[L:  6, P: 15]      |                newline:                                      '\n'
[L:  7, P:  1]      |                whitespace:                                   '    '
[L:  7, P:  5]      |        unparsable:                                           !! Expected: 'Nothing...'
[L:  7, P:  5]      |            raw:                                              'NULL_IF'
[L:  7, P: 12]      |            whitespace:                                       ' '
[L:  7, P: 13]      |            raw:                                              '='
[L:  7, P: 14]      |            whitespace:                                       ' '
[L:  7, P: 15]      |            bracketed:
[L:  7, P: 15]      |                start_bracket:                                '('
[L:  7, P: 16]      |                end_bracket:                                  ')'
[L:  7, P: 17]      |    statement_terminator:                                     ';'
[L:  7, P: 18]      |    newline:                                                  '\n'
[L:  8, P:  1]      |    newline:                                                  '\n'
[L:  9, P:  1]      |    statement:
[L:  9, P:  1]      |        create_file_format_segment:
[L:  9, P:  1]      |            keyword:                                          'CREATE'
[L:  9, P:  7]      |            whitespace:                                       ' '
[L:  9, P:  8]      |            keyword:                                          'FILE'
[L:  9, P: 12]      |            whitespace:                                       ' '
[L:  9, P: 13]      |            keyword:                                          'FORMAT'
[L:  9, P: 19]      |            whitespace:                                       ' '
[L:  9, P: 20]      |            keyword:                                          'IF'
[L:  9, P: 22]      |            whitespace:                                       ' '
[L:  9, P: 23]      |            keyword:                                          'NOT'
[L:  9, P: 26]      |            whitespace:                                       ' '
[L:  9, P: 27]      |            keyword:                                          'EXISTS'
[L:  9, P: 33]      |            whitespace:                                       ' '
[L:  9, P: 34]      |            object_reference:
[L:  9, P: 34]      |                naked_identifier:                             'my_file_format'
[L:  9, P: 48]      |            newline:                                          '\n'
[L: 10, P:  1]      |            whitespace:                                       '    '
[L: 10, P:  5]      |            json_file_format_type_parameters:
[L: 10, P:  5]      |                keyword:                                      'TYPE'
[L: 10, P:  9]      |                whitespace:                                   ' '
[L: 10, P: 10]      |                comparison_operator:
[L: 10, P: 10]      |                    raw_comparison_operator:                  '='
[L: 10, P: 11]      |                whitespace:                                   ' '
[L: 10, P: 12]      |                file_type:                                    'JSON'
[L: 10, P: 16]      |                newline:                                      '\n'
[L: 11, P:  1]      |                whitespace:                                   '    '
[L: 11, P:  5]      |        unparsable:                                           !! Expected: 'Nothing...'
[L: 11, P:  5]      |            raw:                                              'NULL_IF'
[L: 11, P: 12]      |            whitespace:                                       ' '
[L: 11, P: 13]      |            raw:                                              '='
[L: 11, P: 14]      |            whitespace:                                       ' '
[L: 11, P: 15]      |            bracketed:
[L: 11, P: 15]      |                start_bracket:                                '('
[L: 11, P: 16]      |                end_bracket:                                  ')'
[L: 11, P: 17]      |    statement_terminator:                                     ';'
[L: 11, P: 18]      |    newline:                                                  '\n'
[L: 12, P:  1]      |    newline:                                                  '\n'
[L: 13, P:  1]      |    statement:
[L: 13, P:  1]      |        create_file_format_segment:
[L: 13, P:  1]      |            keyword:                                          'CREATE'
[L: 13, P:  7]      |            whitespace:                                       ' '
[L: 13, P:  8]      |            keyword:                                          'FILE'
[L: 13, P: 12]      |            whitespace:                                       ' '
[L: 13, P: 13]      |            keyword:                                          'FORMAT'
[L: 13, P: 19]      |            whitespace:                                       ' '
[L: 13, P: 20]      |            keyword:                                          'IF'
[L: 13, P: 22]      |            whitespace:                                       ' '
[L: 13, P: 23]      |            keyword:                                          'NOT'
[L: 13, P: 26]      |            whitespace:                                       ' '
[L: 13, P: 27]      |            keyword:                                          'EXISTS'
[L: 13, P: 33]      |            whitespace:                                       ' '
[L: 13, P: 34]      |            object_reference:
[L: 13, P: 34]      |                naked_identifier:                             'my_file_format'
[L: 13, P: 48]      |            newline:                                          '\n'
[L: 14, P:  1]      |            whitespace:                                       '    '
[L: 14, P:  5]      |            csv_file_format_type_parameters:
[L: 14, P:  5]      |                keyword:                                      'TYPE'
[L: 14, P:  9]      |                whitespace:                                   ' '
[L: 14, P: 10]      |                comparison_operator:
[L: 14, P: 10]      |                    raw_comparison_operator:                  '='
[L: 14, P: 11]      |                whitespace:                                   ' '
[L: 14, P: 12]      |                file_type:                                    'CSV'
[L: 14, P: 15]      |                newline:                                      '\n'
[L: 15, P:  1]      |                whitespace:                                   '    '
[L: 15, P:  5]      |        unparsable:                                           !! Expected: 'Nothing...'
[L: 15, P:  5]      |            raw:                                              'ESCAPE_UNENCLOSED_FIELD'
[L: 15, P: 28]      |            whitespace:                                       ' '
[L: 15, P: 29]      |            raw:                                              '='
[L: 15, P: 30]      |            whitespace:                                       ' '
[L: 15, P: 31]      |            raw:                                              "'\\\\'"
[L: 15, P: 35]      |    statement_terminator:                                     ';'
[L: 15, P: 36]      |    newline:                                                  '\n'
[L: 16, P:  1]      |    newline:                                                  '\n'
[L: 17, P:  1]      |    statement:
[L: 17, P:  1]      |        create_file_format_segment:
[L: 17, P:  1]      |            keyword:                                          'CREATE'
[L: 17, P:  7]      |            whitespace:                                       ' '
[L: 17, P:  8]      |            keyword:                                          'FILE'
[L: 17, P: 12]      |            whitespace:                                       ' '
[L: 17, P: 13]      |            keyword:                                          'FORMAT'
[L: 17, P: 19]      |            whitespace:                                       ' '
[L: 17, P: 20]      |            keyword:                                          'IF'
[L: 17, P: 22]      |            whitespace:                                       ' '
[L: 17, P: 23]      |            keyword:                                          'NOT'
[L: 17, P: 26]      |            whitespace:                                       ' '
[L: 17, P: 27]      |            keyword:                                          'EXISTS'
[L: 17, P: 33]      |            whitespace:                                       ' '
[L: 17, P: 34]      |            object_reference:
[L: 17, P: 34]      |                naked_identifier:                             'my_file_format'
[L: 17, P: 48]      |            newline:                                          '\n'
[L: 18, P:  1]      |            whitespace:                                       '    '
[L: 18, P:  5]      |            parquet_file_format_type_parameters:
[L: 18, P:  5]      |                keyword:                                      'TYPE'
[L: 18, P:  9]      |                whitespace:                                   ' '
[L: 18, P: 10]      |                comparison_operator:
[L: 18, P: 10]      |                    raw_comparison_operator:                  '='
[L: 18, P: 11]      |                whitespace:                                   ' '
[L: 18, P: 12]      |                file_type:                                    'PARQUET'
[L: 18, P: 19]      |                newline:                                      '\n'
[L: 19, P:  1]      |                whitespace:                                   '\t'
[L: 19, P:  2]      |        unparsable:                                           !! Expected: 'Nothing...'
[L: 19, P:  2]      |            raw:                                              'BINARY_AS_TEXT'
[L: 19, P: 16]      |            whitespace:                                       ' '
[L: 19, P: 17]      |            raw:                                              '='
[L: 19, P: 18]      |            whitespace:                                       ' '
[L: 19, P: 19]      |            raw:                                              'TRUE'
[L: 19, P: 23]      |    statement_terminator:                                     ';'
[L: 19, P: 24]      |    newline:                                                  '\n'
[L: 20, P:  1]      |    newline:                                                  '\n'
[L: 21, P:  1]      |    statement:
[L: 21, P:  1]      |        unparsable:                                           !! Expected: 'StatementSegment'
[L: 21, P:  1]      |            raw:                                              'ALTER'
[L: 21, P:  6]      |            whitespace:                                       ' '
[L: 21, P:  7]      |            raw:                                              'FILE'
[L: 21, P: 11]      |            whitespace:                                       ' '
[L: 21, P: 12]      |            raw:                                              'FORMAT'
[L: 21, P: 18]      |            whitespace:                                       ' '
[L: 21, P: 19]      |            raw:                                              'IF'
[L: 21, P: 21]      |            whitespace:                                       ' '
[L: 21, P: 22]      |            raw:                                              'EXISTS'
[L: 21, P: 28]      |            whitespace:                                       ' '
[L: 21, P: 29]      |            raw:                                              'my_file_format'
[L: 21, P: 43]      |            whitespace:                                       ' '
[L: 21, P: 44]      |            raw:                                              'SET'
[L: 21, P: 47]      |            newline:                                          '\n'
[L: 22, P:  1]      |            whitespace:                                       '    '
[L: 22, P:  5]      |            raw:                                              'STRIP_OUTER_ARRAY'
[L: 22, P: 22]      |            whitespace:                                       ' '
[L: 22, P: 23]      |            raw:                                              '='
[L: 22, P: 24]      |            whitespace:                                       ' '
[L: 22, P: 25]      |            raw:                                              'TRUE'
[L: 22, P: 29]      |    newline:                                                  '\n'
[L: 23, P:  1]      |    statement_terminator:                                     ';'

==== parsing violations ====
L:   3 | P:   2 |  PRS | Line 3, Position 2: Found unparsable section: "ESCAPE = '\\\\'"
L:   7 | P:   5 |  PRS | Line 7, Position 5: Found unparsable section: 'NULL_IF = ()'
L:  11 | P:   5 |  PRS | Line 11, Position 5: Found unparsable section: 'NULL_IF = ()'
L:  15 | P:   5 |  PRS | Line 15, Position 5: Found unparsable section:
                       | "ESCAPE_UNENCLOSED_FIELD = '\\\\'"
L:  19 | P:   2 |  PRS | Line 19, Position 2: Found unparsable section: 'BINARY_AS_TEXT =
                       | TRUE'
L:  21 | P:   1 |  PRS | Line 21, Position 1: Found unparsable section: 'ALTER FILE FORMAT
                       | IF EXISTS my_file_form...'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
```

### How to reproduce

1. Write an file.sql:
```
CREATE FILE FORMAT IF NOT EXISTS my_file_format
    TYPE = CSV
	ESCAPE = '\\';

CREATE FILE FORMAT IF NOT EXISTS my_file_format
    TYPE = CSV
    NULL_IF = ();

CREATE FILE FORMAT IF NOT EXISTS my_file_format
    TYPE = JSON
    NULL_IF = ();

CREATE FILE FORMAT IF NOT EXISTS my_file_format
    TYPE = CSV
    ESCAPE_UNENCLOSED_FIELD = '\\';

CREATE FILE FORMAT IF NOT EXISTS my_file_format
    TYPE = PARQUET
	BINARY_AS_TEXT = TRUE;

ALTER FILE FORMAT IF EXISTS my_file_format SET
    STRIP_OUTER_ARRAY = TRUE
;
```
2.  Run the sqlfluff command `sqlfluff parse file.sql --dialect snowflake`

### Dialect

snowflake

### Version

sqlfluff 1.2.1
Python 3.9.12

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Gal40n04ek**
### Brief summary of the change made
fixes #3773

### Are there any other side effects of this change that we should be aware of?
regenerated `snowflake_create_file_format` fixture
created new `snowflake_alter_file_format` fixture

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 1]
```diff
@@ -3134,7 +3174,10 @@ class CsvFileFormatTypeParameters(BaseSegment):
         Sequence(
             "NULL_IF",
             Ref("EqualsSegment"),
-            Bracketed(Delimited(Ref("QuotedLiteralSegment"))),
+            OneOf(
+                Bracketed(),
+                Bracketed(Delimited(Ref("QuotedLiteralSegment")))
+            ),
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_snowflake.py:3180]: Would this have worked instead?
```suggestion
            Bracketed(Delimited(Ref("QuotedLiteralSegment")), optional=True),
```

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 2]
```diff
@@ -3218,7 +3261,10 @@ class JsonFileFormatTypeParameters(BaseSegment):
         Sequence(
             "NULL_IF",
             Ref("EqualsSegment"),
-            Bracketed(Delimited(Ref("QuotedLiteralSegment"))),
+            OneOf(
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_snowflake.py:3264]: Same question as above.

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 3]
```diff
@@ -3357,6 +3403,7 @@ class ParquetFileFormatTypeParameters(BaseSegment):
         Sequence(
             "SNAPPY_COMPRESSION", Ref("EqualsSegment"), Ref("BooleanLiteralGrammar")
         ),
+        Sequence("BINARY_AS_TEXT", Ref("EqualsSegment"), Ref("BooleanLiteralGrammar")),
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_snowflake.py:3406]: I wonder if all the ones like this:
```
        Sequence("...something...", Ref("EqualsSegment"), Ref("BooleanLiteralGrammar")),
```

could be combined into something like:
```
        Sequence(
            OneOf("SNAPPY_COMPRESSION", "BINARY_AS_TEXT", ...),
            Ref("EqualsSegment"),
            Ref("BooleanLiteralGrammar")),
```

No real concern, just seems tidier.

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 1 above)
Comment (Gal40n04ek) [reply src/sqlfluff/dialects/dialect_snowflake.py:3180]: Actually, yes, it works but with small changes: 
`Bracketed(Delimited(Ref("QuotedLiteralSegment"), optional=True))`
I'll make these changes  now

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 3 above)
Comment (Gal40n04ek) [reply src/sqlfluff/dialects/dialect_snowflake.py:3406]: Yes, it works. I can make these changes if you'd like.

Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_snowflake.py:3401]: Sure, if you don't mind.

**Comment by barrywhart** [comment]
Can you please run "Black" on the code? The code linting check is failing.

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 3 above)
Comment (Gal40n04ek) [reply src/sqlfluff/dialects/dialect_snowflake.py]: Done! Also for the similar statements in all file format types

**Comment by Gal40n04ek** [comment]
> Can you please run "Black" on the code? The code linting check is failing.

Done and fixed!

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 4]
```diff
@@ -3407,20 +3416,16 @@ class XmlFileFormatTypeParameters(BaseSegment):
...
+            OneOf(
+                "IGNORE_UTF8_ERRORS",
+                "PRESERVE_SPACE",
+                "STRIP_OUTER_ELEMENT",
+                "DISABLE_SNOWFLAKE_DATA",
+                "DISABLE_AUTO_CONVERT",
+                "SKIP_BYTE_ORDER_MARK",
+            ),
+            Ref("EqualsSegment"),
+            Ref("BooleanLiteralGrammar"),
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_snowflake.py:3428]: Thanks for doing this! It seems cleaner to me. Possibly faster too -- I was thinking the parser may waste time on partial matches (up to the equals sign) on all the individual ones in the old code. 👍

**Review by barrywhart — APPROVED** [review]
Looks good, thank you!!

**Pull request merged**